src/BCFSerialize.cc

```diff
@@ -0,0 +1,254 @@
+    }
+
+    /*
+      Write the BCF record directly to a memory location. Return how much
+      space was used.
+
+      Note: the code is adapted from the bcf_write routine in htslib/vcf.c.
+      The original prototype is:
+          int bcf_write(htsFile *hfp, const bcf_hdr_t *h, bcf1_t *v)
+      The original routine writes to a file, not to memory.
+    */
+    void bcf_write_to_mem(bcf1_t *v, int reclen, char *addr) {
+        int loc = 0;
+        uint32_t x[8];
+        assert(sizeof(x) == 32);
+
+        x[0] = v->shared.l + 24; // to include six 32-bit integers
+        x[1] = v->indiv.l;
+        memcpy(x + 2, v, 16);
+        x[6] = (uint32_t)v->n_allele<<16 | v->n_info;
+        x[7] = (uint32_t)v->n_fmt<<24 | v->n_sample;
+
+        memcpy(&addr[loc], (char*)x, sizeof(x));
+        loc += sizeof(x);
+        memcpy(&addr[loc], v->shared.s, v->shared.l);
+        loc += v->shared.l;
+        memcpy(&addr[loc], v->indiv.s, v->indiv.l);
+        loc += v->indiv.l;
+
+        assert(loc == reclen);
+    }
+
+
+    /*
+      Read a BCF record from memory, return the length of the packed record in RAM.
+
+      Note: the code is adapted from the bcf_read1_core routine in htslib/vcf.c.
+      The original prototype is:
+          int bcf_read1_core(BGZF *fp, bcf1_t *v)
+      The original routine reads from a file, not from memory.
+    */
+     int bcf_read_from_mem(const char *addr, bcf1_t *v) {
+        int loc = 0;
+        uint32_t x[8];
+        memcpy(x, &addr[loc], 32);
+        loc += 32;
+
+        assert(x[0] > 0);
+        x[0] -= 24; // to exclude six 32-bit integers
+        ks_resize(&v->shared, x[0]);
+        ks_resize(&v->indiv, x[1]);
+        memcpy(v, (char*)&x[2], 16);
+        v->n_allele = x[6]>>16; v->n_info = x[6]&0xffff;
+        v->n_fmt = x[7]>>24; v->n_sample = x[7]&0xffffff;
+        v->shared.l = x[0], v->indiv.l = x[1];
+
+        // silent fix of broken BCFs produced by earlier versions of
+        // bcf_subset, prior to and including bd6ed8b4
+        if ( (!v->indiv.l || !v->n_sample) && v->n_fmt ) v->n_fmt = 0;
+
+        memcpy(v->shared.s, &addr[loc], v->shared.l);
+        loc += v->shared.l;
+        memcpy(v->indiv.s, &addr[loc], v->indiv.l);
+        loc += v->indiv.l;
+
+        return loc;
+    }
+
+
+    int BCFWriter::INIT_SIZE = 8 * 1024;
+    int BCFWriter::SIZE_MULTIPLIER = 2;
+    int BCFWriter::MAX_BUF_SIZE = 16 * 1024 * 1024;
+
+    BCFWriter::BCFWriter() {}
+
+    Status BCFWriter::Open(unique_ptr<BCFWriter>& ans) {
+        ans.reset(new BCFWriter);
+        ans->valid_bytes_ = 0;
+        ans->buf_.reserve(INIT_SIZE);
+        return Status::OK();
+    }
+
+    BCFWriter::~BCFWriter() {
+        buf_.clear();
+        valid_bytes_ = 0;
+    }
+
+    Status BCFWriter::write(bcf1_t* x) {
+        int reclen = bcf_calc_packed_len(x);
+
+        // Note: allocation on the stack. This works well if
+        // the memory allocated is no larger than a few KiB.
```

Comment:
This is cool but perhaps premature. BCF records can be arbitrarily large in general because they can hold any number of samples. For now can we just use a vanilla heap buffer that gets realloc'd when needed (and freed in the destructor :)